api/python/cellxgene_census/src/cellxgene_census/experimental/pytorch.py

```diff
@@ -0,0 +1,437 @@
+import logging
+import os
+import sys
+from datetime import timedelta
+from time import time
+from typing import Any, Dict, Iterator, Optional, Sequence, Tuple
+
+import numpy as np
+import pandas as pd
+import pyarrow as pa
+import scipy
+import somacore
+import tiledbsoma as soma
+import torch
+import torchdata.datapipes.iter as pipes
+from attr import attrs
+from scipy import sparse
+from sklearn.preprocessing import LabelEncoder
+from somacore.query import _fast_csr
+from torch import Tensor
+from torch.utils.data import DataLoader
+from torch.utils.data.dataset import Dataset
+
+import cellxgene_census
+from cellxgene_census._open import _build_soma_tiledb_context
+
+ObsDatum = Tuple[Tensor, torch.sparse_coo_tensor]
+
+Encoders = Dict[str, LabelEncoder]
+
+DEFAULT_BUFFER_BYTES = 1024**3
+
+DEFAULT_WORKER_TIMEOUT = 120
+
+pytorch_logger = logging.getLogger("cellxgene_census.experimental.pytorch")
+pytorch_logger.setLevel(logging.INFO)
+
+
+@attrs
+class Stats:
+    n_obs: int = 0
+    """The total number of obs rows retrieved"""
+
+    nnz: int = 0
+    """The total number of values retrieved"""
+
+    elapsed: int = 0
+    """The total elapsed time in seconds for retrieving all batches"""
+
+    n_soma_batches: int = 0
+    """The number of batches retrieved"""
+
+    def __str__(self) -> str:
+        return (
+            f"n_soma_batches={self.n_soma_batches}, n_obs={self.n_obs}, nnz={self.nnz}, "
+            f"elapsed={timedelta(seconds=self.elapsed)}"
+        )
+
+
+def _open_experiment(
+    uri: str, aws_region: Optional[str] = None, buffer_bytes: int = DEFAULT_BUFFER_BYTES
+) -> soma.Experiment:
+    context = _build_soma_tiledb_context(aws_region).replace(
+        tiledb_config={
+            "py.init_buffer_bytes": buffer_bytes,
+            "soma.init_buffer_bytes": buffer_bytes,
+        }
+    )
+    return soma.Experiment.open(uri, context=context)
+
+
+class _ObsAndXIterator(Iterator[ObsDatum]):
+    """
```

Comment:
Probably not critical for an experimental module, but this docstring should use [Google-style guidelines](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html), which will also look neater and generate a proper docsite page.

Reply:
That reminds me, I need to add a "documentation" issue. Will clean it all up there!